### PR TITLE
Flower spread ABM: Optimise 

### PR DESCRIPTION
**Pull request by paramat**
Flower spread ABM: Optimise

Match maximum spread density to maximum mapgen density for flowers.
Place 3 flora nodes at once instead of 1.
Change ABM chance value to 300 to match previous spread rate.
ABM becomes 3 times less intensive.
///////////

screenshot_20180121_152516

^ Testing with interval 1 and starting from 1 flower, explosive spread

The 'flower spread' ABM in the flowers mod is surprisingly intensive according to profiling, see https://github.com/minetest/minetest_game/issues/2022#issuecomment-358649645
The ABM is intensive because for each ABM 'action' on one plant node it:
* Does 'find nodes in area' for a 9x9x9 node volume, counting flora nodes, if there are more than 3 no spread occurs.
* Does 'find nodes in area under air' for the same volume to find surface nodes to possibly spead to.
* If conditions are met only 1 new plant is placed, after all that.

The idea of this PR is to place 3 new plants at this point instead of just 1, greatly increasing the exponential rate of spread (quadrupling instead of doubling), which will allow us to reduce the 'chance of action' of this ABM. I have reduced the chance by a factor of 3, the ABM becomes 3 times less intensive as well as making spread slightly faster.

The other change is detailed below in https://github.com/minetest/minetest_game/pull/2016#issuecomment-359195056 Flora spreads until it reaches a maximum density, the density was out of date (unchanged since mgv6-only days) and far too low, this PR raises it to match maximum mapgen density of flora.

With slightly faster spread and higher maximum density farming flowers becomes faster and requires less farmed area.

//////////////
For a possible later PR:
I am also considering the idea to restrict spread to surface nodes identical to the surface node of the parent plant, this will stop plants spreading into biomes they are not native to. Let me know if this will break any mods or worlds.

### CODE REVIEW

File: mods/flowers/init.lua [hunk 1]
```diff
@@ -145,26 +145,36 @@ function flowers.flower_spread(pos, node)
...
-				minetest.get_node(seedling).name == "default:desert_sand" then
-			return
+	local num_soils = #soils
+	if num_soils > 0 then
+		-- Get 2 random soil nodes from area
+		local soil_pos = {}
+		soil_pos[1] = soils[math.random(num_soils)]
+		soil_pos[2] = soils[math.random(num_soils)]
+		while soil_pos[2] == soil_pos[1] do
+			soil_pos[2] = soils[math.random(num_soils)]
```
Comment (SmallJoker) [inline mods/flowers/init.lua]: Possible infinite loop if there's only one soil.

Comment (paramat) [reply mods/flowers/init.lua]: Well spotted.

**Comment by paramat** [comment]
A further change needed: The max density of 4 per 9x9 area is out of date, flowers can be at a higher density. I'll find the highest density and reset the maximum. This will allow more flowers to be placed at once (maybe 3), making the optimisation more effective.

**Comment by paramat** [comment]
```
local function register_flower(seed, name)
	minetest.register_decoration({
		deco_type = "simple",
		place_on = {"default:dirt_with_grass"},
		sidelen = 16,
		noise_params = {
			offset = -0.02,
			scale = 0.04,
			spread = {x = 200, y = 200, z = 200},
			seed = seed,
			octaves = 3,
			persist = 0.6
		},
		biomes = {"grassland", "deciduous_forest", "floatland_grassland"},
		y_min = 1,
		y_max = 31000,
		decoration = "flowers:"..name,
	})
end
```
Typical maximum density for one flower type is therefore -0.02 + 0.04 = 0.02.
8 flowers: 0.02 * 8 = 0.16 flowers per node.
Over 9 ^ 2 node area: 0.16 * 9 ^ 2 = 12.96 roughly 13.

**Comment by paramat** [comment]
Updated as detailed, in testing.

File: mods/flowers/init.lua (hunk 1 above)
Comment (paramat) [reply mods/flowers/init.lua]: Avoided this by simplifying and not worrying about duplicated positions.

**Comment by paramat** [comment]
Profiled ABM time per server step (90ms) in grassland with dense grasses and mid-density flowers.
MTG master: Max 6.5ms Av 3.5ms.
This PR: Max 4.1ms Av 1.4ms.

Still the most intensive ABM but only by a factor of 1.4 over the next (by averages).
Average is almost down to 1/3rd.

File: mods/flowers/init.lua (hunk 1 above)
Comment (paramat) [reply mods/flowers/init.lua]: Duplicated positions are unlikely with the typical large areas of soils, avoiding them requires lots of additional code.

File: mods/flowers/init.lua [hunk 2]
```diff
@@ -139,32 +139,36 @@ function flowers.flower_spread(pos, node)
...
+	local num_soils = #soils
+	if num_soils >= 1 then
+		for si = 1, math.min(3, num_soils) do
+			local soil = soils[math.random(num_soils)]
+			local soil_above = {x = soil.x, y = soil.y + 1, z = soil.z}
+			light = minetest.get_node_light(soil_above)
+			if not light or light < 13 or
+					-- Desert sand is in the soil group
+					minetest.get_node(soil).name == "default:desert_sand" then
+				return
```
Comment (SmallJoker) [inline mods/flowers/init.lua]: Don't return here. Leave this case empty and set the node in `else`

Comment (paramat) [reply mods/flowers/init.lua]: Ok.

Comment (paramat) [reply mods/flowers/init.lua]: Ugh a return in a loop, sorry, big mistake.

Comment (paramat) [reply mods/flowers/init.lua]: Fixed, testing.

**Comment by paramat** [comment]
Updated profiling shows no change:

Profiled ABM time per server step (90ms) in grassland with dense grasses and mid-density flowers.
MTG master: Max 6.5ms Av 3.5ms.
This PR: Max 4.4ms Av 1.5ms.
Still the most intensive ABM but only by a factor of 1.5 over the next (by averages).
Average is almost down to 1/3rd.

**Comment by paramat** [comment]
I'm not going to accept this ABM unchanged, the action alone can take up to 9ms, 10% of a server step, it's also insane as it is (see first post and the issue) and i would prefer it removed completely and we use seeds.
However that will likely not be accepted so i will compromise and optimise it. There is no harm in placing 3 instead of 1 at a time and this makes it 3 times less intensive as well as increasing the speed of farming flowers.

**Comment by benrob0329** [comment]
I'd be fine with using seeds.

**Comment by paramat** [comment]
I suggested seeds for flowers a year or so ago and most MTG devs were opposed =)

**Comment by jastevenson303** [comment]
Flowers and grass are about the only dynamic things in MTG currently.

**Comment by paramat** [comment]
WIP realised that the chance can only be halved to match spread rate.

**Comment by paramat** [comment]
Time to spread from 1 plant to 128:

Interval 1

PR placing 1 and with chance 96 (previous behaviour but with PR maximum density limit)
16m05s

PR with chance 288
11m59s

So the PR results in faster spread while being 3 times less intensive.

**Comment by paramat** [comment]
There have been requests for faster flower farming, but should the spread rate be equal to before? I can tune it to be.

**Comment by JurajVajda** [comment]
Since you already made the performance improvement using the same spreading speed i would say keep the spreading speed as it was before - not that fast anyway. Other option could be slower spreading with addition of bone meal in the near future.

**Comment by Ezhh** [comment]
I would prefer no big increase (a little is certainly fine). People also have said they'd like fertiliser of some kind added, which could be reserved for nice speed increases.

**Comment by paramat** [comment]
Ok good a decision is made, will tune.

**Comment by paramat** [comment]
After lengthy tuning.

Time to spread from 1 plant to 128 in empty flat grassland with interval 1s:
PR with chance 300
13m29s
(MTG master is 16m05s)
19% faster spread.

I tried chance values over 300 to get a closer match but these made the first spread action far too rare, i had to abort the tests due to nothing happening. Close enough.

**Comment by JurajVajda** [comment]
Sounds good, thanks!

**Comment by paramat** [comment]
Testing has shown density becomes higher than 13 per 9x9, will tune.

**Comment by Ezhh** [comment]
9*9 areas for testing show far higher than the anticipated average spread:
ekkflowers
We're getting up to 23 flowers per 9*9, which I feel is far too dense.

**Comment by paramat** [comment]
Average maximum density per 9x9 seemed to be 21.
So i multiply the limit of 11 by 13/21, result = 7.
I tried 7 here:
```
+	-- Maximum mapgen flower density is 13 per 9x9 area, the limit below is
+	-- tuned by trial and error to result in the same maximum density.
+	-- Setting this limit theoretically does not work in practice.
+	if #minetest.find_nodes_in_area(pos0, pos1, "group:flora") > 7 then
 		return
 	end
```
Ran spread until placing almost stopped (max density).
Then added code to print number per 9x9 (already calculated in the code) and print a running average of that.
Result:
```
flora in 9x9 11
av flora in 9x9 13.617611741161
flora in 9x9 12
av flora in 9x9 13.617395944504
flora in 9x9 11
av flora in 9x9 13.617046818727
flora in 9x9 11
av flora in 9x9 13.616697786076
flora in 9x9 12
av flora in 9x9 13.616482197626
flora in 9x9 13
av flora in 9x9 13.6164
flora in 9x9 18
av flora in 9x9 13.61698440208
flora in 9x9 15
av flora in 9x9 13.617168754999
flora in 9x9 13
av flora in 9x9 13.617086498734
flora in 9x9 12
av flora in 9x9 13.616871002132
```
Close, will also try a limit of 6 but expect that will be too low.

**Comment by paramat** [comment]
The same with a coded limit of 6:
```
flora in 9x9 7
av flora in 9x9 12.024097874444
flora in 9x9 12
av flora in 9x9 12.024094896824
flora in 9x9 14
av flora in 9x9 12.024339016555
flora in 9x9 11
av flora in 9x9 12.024212476838
flora in 9x9 9
av flora in 9x9 12.023838932806
flora in 9x9 13
av flora in 9x9 12.02395949117
flora in 9x9 11
av flora in 9x9 12.023833045196
flora in 9x9 14
av flora in 9x9 12.024077046549
flora in 9x9 10
av flora in 9x9 12.023827160494
flora in 9x9 15
av flora in 9x9 12.024194543883
```
13.6 is closer to 13 than 12, so 7 is best, and the PR is already7.

**Comment by Ezhh** [comment]
Much better now. Thanks for adjusting. :+1:

**Comment by paramat** [comment]
Updated comment.

**Comment by paramat** [comment]
a587972010d8ecf83f10010ada91d43db06f976c

**Pull request closed without merging**